### PR TITLE
New device store implementation

### PR DESCRIPTION
**Pull request by Andrea-Campanella**
This Pull Request moves the the `config` subsystem to the latest version fo `device/store.go` to depend upon `topo`. Also introduces testing through a mock of the store and removes one level of events generation by removing the `topoEvent` and using directly a channel of `device.ListResponse` that contain all the info related to the event that are used by the `synchroniser`. No changes were introduced in either tests or integration tests except for the mocks, this functionality wise this patch is identical to previous versions.

### CODE REVIEW

File: pkg/northbound/gnmi/get_test.go [hunk 1]
```diff
@@ -58,7 +62,9 @@ func Test_getWithPrefixNoOtherPathsNoTarget(t *testing.T) {
 
 // Test_getNoPathElems tests for  Paths with no elements - should treat it like /
 func Test_getNoPathElems(t *testing.T) {
-	server, _ := setUp()
+	server, _, mockStore := setUp(t)
+	mockStore.EXPECT().Get(gomock.Any()).Return(nil, status.Error(codes.NotFound, "device not found"))
```
Comment (SeanCondon) [inline pkg/northbound/gnmi/get_test.go]: you could use `.Times(2)` here I think

File: pkg/northbound/gnmi/set.go [hunk 2]
```diff
@@ -115,15 +115,15 @@ func (s *Server) Set(ctx context.Context, req *gnmi.SetRequest) (*gnmi.SetRespon
...
 		err := validateChange(target, version, deviceType, updates, targetRemoves[target])
 		if err != nil {
 			return nil, err
 		}
 		delete(targetRemovesTmp, target)
-		if _, ok := mgr.DeviceStore.Cache[device.ID(target)]; !ok {
+		_, errDevice := manager.GetManager().DeviceStore.Get(device.ID(target))
+		if errDevice != nil && status.Convert(errDevice).Code() == codes.NotFound {
 			disconnectedDevices = append(disconnectedDevices, target)
 		}
```
Comment (SeanCondon) [inline pkg/northbound/gnmi/set.go]: Maybe should handle the `err` for other codes too

File: pkg/northbound/gnmi/set.go [hunk 3]
```diff
@@ -133,7 +133,8 @@ func (s *Server) Set(ctx context.Context, req *gnmi.SetRequest) (*gnmi.SetRespon
 		if err != nil {
 			return nil, err
 		}
-		if _, ok := mgr.DeviceStore.Cache[device.ID(target)]; !ok {
+		_, errDevice := manager.GetManager().DeviceStore.Get(device.ID(target))
+		if errDevice != nil && status.Convert(errDevice).Code() == codes.NotFound {
 			disconnectedDevices = append(disconnectedDevices, target)
 		}
```
Comment (SeanCondon) [inline pkg/northbound/gnmi/set.go]: same

File: pkg/northbound/gnmi/set_test.go [hunk 4]
```diff
@@ -248,7 +255,9 @@ func Test_do2SetsOnSameTarget(t *testing.T) {
 // Test_do2SetsOnDiffTargets shows how paths on multiple targets can be Set at
 // same time
 func Test_do2SetsOnDiffTargets(t *testing.T) {
-	server, _ := setUp()
+	server, _, mockStore := setUp(t)
+	mockStore.EXPECT().Get(gomock.Any()).Return(nil, status.Error(codes.NotFound, "device not found"))
+	mockStore.EXPECT().Get(gomock.Any()).Return(nil, status.Error(codes.NotFound, "device not found"))
```
Comment (SeanCondon) [inline pkg/northbound/gnmi/set_test.go]: Use `.Times(2)`

File: pkg/northbound/gnmi/set_test.go [hunk 5]
```diff
@@ -283,7 +292,8 @@ func Test_do2SetsOnDiffTargets(t *testing.T) {
 
 	extensionDeviceState := setResponse.Extension[1].GetRegisteredExt()
 	assert.Equal(t, extensionDeviceState.Id.String(), strconv.Itoa(GnmiExtensionDevicesNotConnected))
-	assert.Equal(t, string(extensionDeviceState.Msg), "localhost-1,localhost-2")
+	assert.Assert(t, strings.Contains(string(extensionDeviceState.Msg), "localhost-1"))
```
Comment (SeanCondon) [inline pkg/northbound/gnmi/set_test.go]: nice - I had this in my patch too

File: pkg/northbound/gnmi/set_test.go [hunk 6]
```diff
@@ -295,7 +305,9 @@ func Test_do2SetsOnDiffTargets(t *testing.T) {
 // Test_do2SetsOnOneTargetOneOnDiffTarget shows how multiple paths on multiple
 // targets can be Set at same time
 func Test_do2SetsOnOneTargetOneOnDiffTarget(t *testing.T) {
-	server, _ := setUp()
+	server, _, mockStore := setUp(t)
+	mockStore.EXPECT().Get(gomock.Any()).Return(nil, status.Error(codes.NotFound, "device not found"))
+	mockStore.EXPECT().Get(gomock.Any()).Return(nil, status.Error(codes.NotFound, "device not found"))
```
Comment (SeanCondon) [inline pkg/northbound/gnmi/set_test.go]: Use `.Times(2)`

File: pkg/northbound/gnmi/set_test.go [hunk 7]
```diff
@@ -353,7 +365,9 @@ func Test_do2SetsOnOneTargetOneOnDiffTarget(t *testing.T) {
 // Test_doDuplicateSetSingleTarget shows how duplicate combineation of paths on
 // a single target fails
 func Test_doDuplicateSetSingleTarget(t *testing.T) {
-	server, _ := setUp()
+	server, _, mockStore := setUp(t)
+	mockStore.EXPECT().Get(gomock.Any()).Return(nil, status.Error(codes.NotFound, "device not found"))
+	mockStore.EXPECT().Get(gomock.Any()).Return(nil, status.Error(codes.NotFound, "device not found"))
```
Comment (SeanCondon) [inline pkg/northbound/gnmi/set_test.go]: same

File: pkg/northbound/gnmi/subscribe.go [hunk 8]
```diff
@@ -326,7 +326,8 @@ func buildSubscribeResponse(notification *gnmi.Notification, target string) *gnm
 	response := &gnmi.SubscribeResponse{
 		Response: responseUpdate,
 	}
-	if _, ok := manager.GetManager().DeviceStore.Cache[device.ID(target)]; !ok {
+	_, err := manager.GetManager().DeviceStore.Get(device.ID(target))
+	if err != nil && status.Convert(err).Code() == codes.NotFound {
```
Comment (SeanCondon) [inline pkg/northbound/gnmi/subscribe.go]: maybe handle the `err` for other codes too

File: pkg/northbound/gnmi/subscribe_test.go [hunk 9]
```diff
@@ -119,7 +121,9 @@ func Test_SubscribeLeafOnce(t *testing.T) {
 
 // Test_SubscribeLeafDelete tests subscribing with mode STREAM and then issuing a set request with updates for that path
 func Test_SubscribeLeafStream(t *testing.T) {
-	server, mgr := setUp()
+	server, mgr, mockStore := setUp(t)
+	mockStore.EXPECT().Get(gomock.Any()).Return(nil, status.Error(codes.NotFound, "device not found"))
```
Comment (SeanCondon) [inline pkg/northbound/gnmi/subscribe_test.go]: again `.Times(2)`

File: pkg/manager/manager.go [hunk 10]
```diff
@@ -38,11 +37,10 @@ var mgr Manager
 type Manager struct {
 	ConfigStore             *store.ConfigurationStore
 	ChangeStore             *store.ChangeStore
-	DeviceStore             *topocache.DeviceStore
-	NewDeviceStore          devicetopo.Store
+	DeviceStore             devicetopo.Store
 	NetworkStore            *store.NetworkStore
 	ModelRegistry           *modelregistry.ModelRegistry
-	TopoChannel             chan events.TopoEvent
+	TopoChannel             chan *devicepb.ListResponse
```
Comment (SeanCondon) [inline pkg/manager/manager.go]: nice - you're taking it straight from the gRPC - no intermediate TopoEvent needed then

File: pkg/northbound/gnmi/get_test.go (hunk 1 above)
Comment (Andrea-Campanella) [reply pkg/northbound/gnmi/get_test.go]: done

File: pkg/northbound/gnmi/set.go (hunk 2 above)
Comment (Andrea-Campanella) [reply pkg/northbound/gnmi/set.go]: added ```else if errDevice != nil {
			//handling gRPC errors
			return nil, err
		}``` to handle any unexpected error, only know thrown by topo is NotFound (see line  149 of service.go under `onos-topo`

File: pkg/northbound/gnmi/set.go (hunk 3 above)
Comment (Andrea-Campanella) [reply pkg/northbound/gnmi/set.go]: same

File: pkg/northbound/gnmi/set_test.go (hunk 4 above)
Comment (Andrea-Campanella) [reply pkg/northbound/gnmi/set_test.go]: done

File: pkg/northbound/gnmi/set_test.go (hunk 6 above)
Comment (Andrea-Campanella) [reply pkg/northbound/gnmi/set_test.go]: done

File: pkg/northbound/gnmi/set_test.go (hunk 7 above)
Comment (Andrea-Campanella) [reply pkg/northbound/gnmi/set_test.go]: done

File: pkg/northbound/gnmi/subscribe.go (hunk 8 above)
Comment (Andrea-Campanella) [reply pkg/northbound/gnmi/subscribe.go]: done

File: pkg/northbound/gnmi/subscribe_test.go (hunk 9 above)
Comment (Andrea-Campanella) [reply pkg/northbound/gnmi/subscribe_test.go]: done

File: pkg/manager/manager.go (hunk 10 above)
Comment (Andrea-Campanella) [reply pkg/manager/manager.go]: yes, one less level of transaction needed. I couple TOPO with CONFIG a bit tighter but it's a tradeoff I'm willing to make for sepmplification's sake

**Pull request merged**